bindings/go/dag/dag.go

```diff
@@ -0,0 +1,328 @@
+// # Modified from https://github.com/kro-run/kro/blob/7e437f2fe159a1e1c59d8eefd2bfa55320df4489/pkg/graph/dag/dag.go under Apache 2.0 License
+//
+// Original License:
+//
+// Copyright 2025 The Kube Resource Orchestrator Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+// We would like to thank the authors of kro for their outstanding work on this code.
+
+package dag
+
+import (
+	"cmp"
+	"fmt"
+	"maps"
+	"slices"
+	"sort"
+	"strings"
+)
+
+var ErrSelfReference = fmt.Errorf("self-references are not allowed")
+
+// Vertex represents a node/vertex in a directed acyclic graph.
+type Vertex[T cmp.Ordered] struct {
```

Comment:
I wonder. Shouldn't we make this concurrent safe with a mutex?

Reply:
I think if we want concurrency safety, it needs to be on the entire map and not just on the vertex. WDYT?

Reply:
Yeah, agreed.